### PR TITLE
metadata: write original md before appended md

### LINKED ISSUE #2878: context metadata order?
Please see the FAQ in our main README.md, then answer the questions below before
submitting your issue.

### What version of gRPC are you using?
v1.12.1

### What version of Go are you using (`go version`)?
1.12.6

### What operating system (Linux, Windows, …) and version?
darwin/amd64

### What did you do?
If possible, provide a recipe for reproducing the error.

```
ctx := metadata.NewOutgoingContext(context.Background(), metadata.Pairs("a", "1"))
ctx = metadata.AppendToOutgoingContext(ctx, "a", "2")
ctx = metadata.AppendToOutgoingContext(ctx, "a", "3")

the outgoing md for a is [1,2,3]
```

but when server received from incoming, the order became [2,3,1].

so, how can i get latest one?

### What did you expect to see?
incoming get [1,2,3], maybe this same as outgoing

createHeaderFields maybe should changed?
https://github.com/grpc/grpc-go/blob/master/internal/transport/http2_client.go#L393

### PR DESCRIPTION
**Pull request by elvizlai**
fix #2878

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
CLA GitHub bot_

**Comment by elvizlai** [comment]
CLA

**Comment by elvizlai** [comment]
CLA

**Comment by elvizlai** [comment]
GAE failed?

**Comment by menghanl** [comment]
Thanks for the fix. The failure is a flaky test.

Can you add a test? Something similar to https://github.com/grpc/grpc-go/blob/ecb921ddb97115e42be19051508c4474c3b387aa/test/end2end_test.go#L3074

**Comment by elvizlai** [comment]
> Thanks for the fix. The failure is a flaky test.
> 
> Can you add a test? Something similar to
> 
> https://github.com/grpc/grpc-go/blob/ecb921ddb97115e42be19051508c4474c3b387aa/test/end2end_test.go#L3074

add TestMetadataOrderUnaryRPC

File: test/end2end_test.go [hunk 1]
```diff
@@ -3116,6 +3116,42 @@ func testMetadataUnaryRPC(t *testing.T, e env) {
...
+	te := newTest(t, e)
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+	tc := testpb.NewTestServiceClient(te.clientConn())
+
+	ctx := metadata.NewOutgoingContext(context.Background(), testMetadata)
+	ctx = metadata.AppendToOutgoingContext(ctx, "key1", "value2")
+	ctx = metadata.AppendToOutgoingContext(ctx, "key1", "value3")
+
+	newMetadata, _ := metadata.FromOutgoingContext(ctx)
```
Comment (menghanl) [inline test/end2end_test.go]: I'm a bit paranoid about this. `FromOutgoingContext` could also return the wrong order.

How about build the expected metadata explicitly:
```go
newMetadata := metadata.Join(testMetadata, metadata.Pairs("key1", "value2","key1", "value3"))
```

Comment (elvizlai) [reply test/end2end_test.go]: Yes, `FromOutgoingContext` is some wrap of `Join`, we can using `Join` directly.

**Pull request merged**